03_letters/jecofang/letter_frequency_test.go

```diff
@@ -0,0 +1,74 @@
+package main
+
+import (
+	"bytes"
+	"strconv"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestMainOutput(t *testing.T) {
+	// Given
+	r := require.New(t)
+	var buf bytes.Buffer
+	out = &buf
+
+	// When
+	main()
+
+	// Then
+	expected := strconv.Quote(`a:2
+b:1
+`)
+	actual := strconv.Quote(buf.String())
+	r.Equalf(expected, actual, "Unexpected output in main()")
+}
+
+var testData = []struct {
+	in      string
+	letters map[rune]int
+	sorted  []string
+}{
+	{"", map[rune]int{}, []string{}},
+	{"aba", map[rune]int{('a'): 2, ('b'): 1}, []string{"a:2", "b:1"}},
+	{"abaAbA", map[rune]int{('a'): 2, ('b'): 2, ('A'): 2}, []string{"A:2", "a:2", "b:2"}},
+	{"ab aAbA", map[rune]int{('a'): 2, ('b'): 2, (' '): 1, ('A'): 2}, []string{" :1", "A:2", "a:2", "b:2"}},
+	{"ab aAbA21",
+		map[rune]int{('a'): 2, ('b'): 2, (' '): 1, ('A'): 2, ('2'): 1, ('1'): 1},
+		[]string{" :1", "1:1", "2:1", "A:2", "a:2", "b:2"}},
+	{"ab aAbA21你 ",
+		map[rune]int{('a'): 2, ('b'): 2, (' '): 2, ('A'): 2, ('2'): 1, ('1'): 1, ('你'): 1},
+		[]string{" :2", "1:1", "2:1", "A:2", "a:2", "b:2", "你:1"}},
+	{"ab,aAbA21你，",
+		map[rune]int{('a'): 2, ('b'): 2, (','): 1, ('A'): 2, ('2'): 1, ('1'): 1, ('你'): 1, ('，'): 1},
+		[]string{",:1", "1:1", "2:1", "A:2", "a:2", "b:2", "你:1", "，:1"}},
+}
+
+func TestLetter(t *testing.T) {
+	// Given
+	r := require.New(t)
```

Comment:
`// Given` doesn't apply here, since `r` is just part of the testing infrastructure, not a precondition for the tests. Consider that removing it and rewriting `r.Equal(…)` as `require.Equal(t, …)` would leave you with exactly the same test.